### ISSUE TITLE
Is there trouble in piper_train.preprocess?

### ISSUE THREAD

**Issue by ican24** [rhasspy/piper#37]
Thank you for this project!
It is so important.
 
I am trying to check and use it, but I met same trouble on 2 machine.
It seems the piper_train.preprocess frozen, when I am trying to run this command.
It is more than 24 hours.
```
~/piper/src/python$ python3 -m piper_train.preprocess \
  --language en-us \
  --input-dir /home/hayq/piper/data/LJSpeech-1.1/ \
  --output-dir /home/hayq/piper/train/ \
  --dataset-format ljspeech \
  --sample-rate 22050
INFO:preprocess:13059 speakers detected
INFO:preprocess:Wrote dataset config
INFO:preprocess:Processing 13100 utterance(s) with 6 worker(s)

```  
The processing seems defuncted.

```
 200794 pts/1    Sl+    0:09 python3 -m piper_train.preprocess --language en-us --input-dir /home/hayq/piper/data/LJSpeech-1.1/ --output-dir /home/hayq/piper/train/ --dataset-format ljspeech --sample-rate 22050
 200806 pts/1    Z+     0:00 [python3] <defunct>
 200809 pts/1    Z+     0:00 [python3] <defunct>
 200812 pts/1    Z+     0:00 [python3] <defunct>
 200815 pts/1    Z+     0:00 [python3] <defunct>
 200816 pts/1    Z+     0:00 [python3] <defunct>
 200817 pts/1    Z+     0:00 [python3] <defunct>

```
Could you help me understand the reason of the trouble.
Thank you in advance!

**Comment by ican24** [comment]
Unfortunately, this script shows most unwanted behavior in computing: it starts but not end with unknown sign.
At least it's required logging feature in "make_dataset" function of preprocess.py (line 158). 
Such way you may waste our time with this software.
 
```
    with open(args.output_dir / "dataset.jsonl", "w", encoding="utf-8") as datasetfile:
        for utt_batch in batched(
            make_dataset(args.input_dir, args.single_speaker, args.speaker_id),
            batch_size,
        ):

```

**Comment by trunglebka** [comment]
> Such way you may waste our time with this software.

Don't make judgements too quickly.
(The only different is sample rate but that does not matter)
![image](https://user-images.githubusercontent.com/16150250/232179592-049a612e-7946-4b00-b20a-147639943146.png)

**Comment by ican24** [comment]
It is a third machine, which I used to test this procedure with original (LJSpeech-1.1) and different datasets even 100-200 records, but the result is a same at 2-nd day: the machines are frozen, no any sign of life.
I need to investigate  "make_dataset" function of preprocess.py to fix this problem. I delayed it waiting author's response, but in this situation I am forced to dig into author's code.

**Comment by ican24** [comment]
The trouble in "batched" function

```
def batched(iterable, n):
    "Batch data into lists of length n. The last batch may be shorter."
    # batched('ABCDEFG', 3) --> ABC DEF G
    if n < 1:
        raise ValueError("n must be at least one")
    it = iter(iterable)
    batch = list(itertools.islice(it, n))
    while batch:
        yield batch
        batch = list(itertools.islice(it, n))


```
The workers are not working.
`utt = queue_out.get()
`
is not processed.
I am tring to fix or change it.

**Comment by trunglebka** [comment]
Not sure what kind of machine you have used, just two recommend:
- run this on linux server with abundant RAM
- monitor your RAM when running to check if it is OOM

**Comment by ican24** [comment]
I am using 3 different machines on Ubuntu and Fedora:
1. Xeon E3-1200 v5/v6 / E3-1500 v5 with 32G memory (Ubuntu 18.04)
2. Intel(R) Core(TM) i5-8400 CPU @ 2.80GH with 16G memory (Ubuntu 22.04)
3. laptop (Fedora)
with standard, default configuration.
It's really the workers are not working.
In this situation I need completely to change preprocess.py or refuse this project.

**Comment by trunglebka** [comment]
Have you monitored your RAM usage?
32GB seems enough but if too many workers were spawned it may cause OOM too.
Computer usually "frozen" only when it is OOM but not full CPU usage

**Comment by ican24** [comment]
Dear @trunglebka, 
How can I check memory?
Please, advise your method.

**Comment by trunglebka** [comment]
Just run the script and monitor with `htop`. 
If memory is full, you can try to reduce number of workers via `--max-workers`

**Comment by trunglebka** [comment]
Just run the script and monitor with `htop`. 
If memory is full, you can try to reduce number of worker via `--max-workers`

**Comment by ican24** [comment]
Thank you!

**Comment by ican24** [comment]
Now I am getting error in piper_train.preprocess script
```
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe.
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe.
Terminating: fork() called from a process already using GNU OpenMP, this is unsafe.
```
The google seaching is useless.
It is not OOM case, becase I use small piece of dataset. Just 50-60 lines.

So It seems a deadlock of exotic codding.

**Comment by trunglebka** [comment]
Can you create new conda env and re-run with that env?

**Comment by ican24** [comment]
It is a complicated issue, because I have a lot of stable installations on the machines.
Never I met trouble with conda and env.
I am so interested to try this projects, but not ready to risk existing installations.
I'll try investigate preprocess procedure to understand what kind of data it generates, then I'll make dataset manually.
Otherwise I will refuse from this project as suspected, unfortunately, because I lose enough time during 3-4 days.

**Comment by ican24** [comment]
The reason is 

`            utt = queue_out.get()
`
the line 166 of preprocess.py

If I mention timeout, it reports that
_queue.Empty

**Comment by methosmythos** [comment]
Hi,

I have the same problem.
Tried in on a Ubuntu i7 machine with 64Gb RAM and python 3.10.6 installed.

`INFO:preprocess:7330 speakers detected
INFO:preprocess:Wrote dataset config
INFO:preprocess:Processing 7333 utterance(s) with 8 worker(s)
`
This never finishes. If I interrupt it with Ctrl+C, i get the following:

`^CTraceback (most recent call last):
  File "/opt/conda/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/opt/conda/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/build/piper/src/python/piper_train/preprocess.py", line 320, in <module>
    main()
  File "/build/piper/src/python/piper_train/preprocess.py", line 164, in main
    utt = queue_out.get()
  File "/opt/conda/lib/python3.8/multiprocessing/queues.py", line 97, in get
    res = self._recv_bytes()
  File "/opt/conda/lib/python3.8/multiprocessing/connection.py", line 216, in recv_bytes
    buf = self._recv_bytes(maxlength)
  File "/opt/conda/lib/python3.8/multiprocessing/connection.py", line 414, in _recv_bytes
    buf = self._recv(4)
  File "/opt/conda/lib/python3.8/multiprocessing/connection.py", line 379, in _recv
    chunk = read(handle, remaining)
KeyboardInterrupt
^CError in atexit._run_exitfuncs:
Traceback (most recent call last):
  File "/opt/conda/lib/python3.8/multiprocessing/util.py", line 300, in _run_finalizers
    finalizer()
  File "/opt/conda/lib/python3.8/multiprocessing/util.py", line 224, in __call__
    res = self._callback(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.8/multiprocessing/queues.py", line 195, in _finalize_join
    thread.join()
  File "/opt/conda/lib/python3.8/threading.py", line 1011, in join
    self._wait_for_tstate_lock()
  File "/opt/conda/lib/python3.8/threading.py", line 1027, in _wait_for_tstate_lock
    elif lock.acquire(block, timeout):
KeyboardInterrupt
`
I also tried it with the docker image and python3 3.8.12 and the same thing happens.
Sadly I'm not proficient in pythong, but it seems like some kind of a deadlock?

Any ideas?

**Comment by ican24** [comment]
Hi,
It seems the multiprocessing module of python was not realized in proper way.
The author uses very complicated way of multiprocessing module.
`
from multiprocessing import JoinableQueue, Process, Queue
`
Now I am trying to fix the problem, but my knowledge of multiprocessing is not so deep.
I am surprised why the author has not responded for 4 days?
Very strange

**Comment by methosmythos** [comment]
Hi,

I've played around a bit and added some logs into the preprocess.py code.
As far as I can tell all process_batch processes are started. At least in my case, the problem occurs at the line 203: `utt.phonemes = phonemize(utt.text, phonemizer)`
I've tried to manually run the test case (listed at: https://pypi.org/project/espeak-phonemizer/): `echo 'This is a test.' | espeak-phonemizer -v en-us`.
This fails with: `Segmentation fault (core dumped)`
So I think the problem (at least in my case) is with the espeak-phonemizer which fails and this probably then has the queue deadlock as a consequence.
The curious thing is that I've tried this on another machine with an older ubuntu and python 3.8.10 and it works (both the espeak-phonemizer and also the piper_train.preprocess).
So maybe it has something to do with this not working in newer python versions?

**Comment by methosmythos** [comment]
Hi,

it seems the issue about espeak-phonemizer is already raised here: https://github.com/rhasspy/espeak-phonemizer/issues/1
As far as I can tell it has to do with the python version.
In my case it seems to work with python 3.8.10, but not with 3.8.12 and higher.

**Comment by ican24** [comment]
@methosmythos , thank you very much!
Everything is an important to fix this problem.

**Comment by trunglebka** [comment]
> Hi,
> 
> it seems the issue about espeak-phonemizer is already raised here: [rhasspy/espeak-phonemizer#1](https://github.com/rhasspy/espeak-phonemizer/issues/1) As far as I can tell it has to do with the python version. In my case it seems to work with python 3.8.10, but not with 3.8.12 and higher.

I'm running it with python 3.10.10

**Comment by ican24** [comment]
It seems the "Segmentation fault (core dumped)" error is avoided, if we install older version of espeak_phonemizer.

`pip install espeak_phonemizer==1.1.0`

**Comment by methosmythos** [comment]
OK, I've just set-up a docker container, based on python 3.8.10 and I still get a segmentation fault for the espeak-phonemizer - so I have no idea, why it's working on the other machine with python 3.8.10 installed, but not on the other inside a docker container with the same python version. So I guess python version is not really the relevant factor here...also, @trunglebka has no problems with 3.10.10.

**Comment by trunglebka** [comment]
this is my conda environment:
```
_libgcc_mutex=0.1=main
_openmp_mutex=5.1=1_gnu
absl-py=1.4.0=pypi_0
aiohttp=3.8.4=pypi_0
aiosignal=1.3.1=pypi_0
appdirs=1.4.4=pypi_0
async-timeout=4.0.2=pypi_0
attrs=22.2.0=pypi_0
audioread=3.0.0=pypi_0
bzip2=1.0.8=h7b6447c_0
ca-certificates=2023.01.10=h06a4308_0
cachetools=5.3.0=pypi_0
certifi=2022.12.7=py310h06a4308_0
cffi=1.15.1=pypi_0
charset-normalizer=3.1.0=pypi_0
coloredlogs=15.0.1=pypi_0
cython=0.29.34=pypi_0
decorator=5.1.1=pypi_0
espeak-phonemizer=1.2.0=pypi_0
flatbuffers=23.3.3=pypi_0
frozenlist=1.3.3=pypi_0
fsspec=2023.4.0=pypi_0
google-auth=2.17.3=pypi_0
google-auth-oauthlib=1.0.0=pypi_0
grpcio=1.53.0=pypi_0
humanfriendly=10.0=pypi_0
idna=3.4=pypi_0
joblib=1.2.0=pypi_0
lazy-loader=0.2=pypi_0
ld_impl_linux-64=2.38=h1181459_1
libffi=3.4.2=h6a678d5_6
libgcc-ng=11.2.0=h1234567_1
libgomp=11.2.0=h1234567_1
librosa=0.10.0.post2=pypi_0
libstdcxx-ng=11.2.0=h1234567_1
libuuid=1.41.5=h5eee18b_0
lightning-utilities=0.8.0=pypi_0
llvmlite=0.39.1=pypi_0
markdown=3.4.3=pypi_0
markupsafe=2.1.2=pypi_0
mpmath=1.3.0=pypi_0
msgpack=1.0.5=pypi_0
multidict=6.0.4=pypi_0
ncurses=6.4=h6a678d5_0
numba=0.56.4=pypi_0
numpy=1.23.5=pypi_0
nvidia-cublas-cu11=11.10.3.66=pypi_0
nvidia-cuda-nvrtc-cu11=11.7.99=pypi_0
nvidia-cuda-runtime-cu11=11.7.99=pypi_0
nvidia-cudnn-cu11=8.5.0.96=pypi_0
oauthlib=3.2.2=pypi_0
onnxruntime=1.14.1=pypi_0
openssl=1.1.1t=h7f8727e_0
packaging=23.1=pypi_0
pip=23.0.1=py310h06a4308_0
pooch=1.6.0=pypi_0
protobuf=4.22.3=pypi_0
pyasn1=0.4.8=pypi_0
pyasn1-modules=0.2.8=pypi_0
pycparser=2.21=pypi_0
pydeprecate=0.3.2=pypi_0
python=3.10.10=h7a1cb2a_2
pytorch-lightning=1.7.7=pypi_0
pyyaml=6.0=pypi_0
readline=8.2=h5eee18b_0
requests=2.28.2=pypi_0
requests-oauthlib=1.3.1=pypi_0
resampy=0.4.2=pypi_0
rsa=4.9=pypi_0
scikit-learn=1.2.2=pypi_0
scipy=1.10.1=pypi_0
setuptools=65.6.3=py310h06a4308_0
six=1.16.0=pypi_0
soundfile=0.12.1=pypi_0
soxr=0.3.5=pypi_0
sqlite=3.41.1=h5eee18b_0
sympy=1.11.1=pypi_0
tensorboard=2.12.1=pypi_0
tensorboard-data-server=0.7.0=pypi_0
tensorboard-plugin-wit=1.8.1=pypi_0
threadpoolctl=3.1.0=pypi_0
tk=8.6.12=h1ccaba5_0
torch=1.13.1=pypi_0
torchmetrics=0.11.4=pypi_0
tqdm=4.65.0=pypi_0
typing-extensions=4.5.0=pypi_0
tzdata=2023c=h04d1e81_0
urllib3=1.26.15=pypi_0
werkzeug=2.2.3=pypi_0
wheel=0.38.4=py310h06a4308_0
xz=5.2.10=h5eee18b_1
yarl=1.8.2=pypi_0
zlib=1.2.13=h5eee18b_0

```

**Comment by alisherski** [comment]
> It seems the "Segmentation fault (core dumped)" error is avoided, if we install older version of espeak_phonemizer.
> 
> `pip install espeak_phonemizer==1.1.0`

Thanks! I encountered the same error and can confirm that downgrading prevents segmentation fault. Launched preprocessing step, waiting for finish.

**Comment by trunglebka** [comment]
I was able to run preprocessing of latest commit on another computer (fedora 37). Just unspecify onnxruntime version (python 3.10):
![image](https://user-images.githubusercontent.com/16150250/232326559-ae0147ef-d062-4549-80f4-53751dfd540b.png)

**Comment by methosmythos** [comment]
I've put some logging info into the espeak phonemizer project and it appears that the breaking change was made in the latest version. An external call to `TextToPhonemes` is used in the latest version (line 101: `self.lib_espeak.espeak_TextToPhonemes(`).
In the older version there is a comment that this call creates a segmentation fault, so a workaround was made.
But in the newest version the call to `TextToPhonemes` is used anyways.
I guess it doesn't generate the segmentation fault on every system, so for now it seems that on those that segmentation fault is generated, the option is to downgrade to a previous version, like @ican24 suggested.

**Comment by ican24** [comment]
Now I can say that the problem was fixed, if 
`src/python/requirements.txt`
is edited such way:

```
cython>=0.29.0,<1
espeak-phonemizer~=1.1.0
librosa>=0.9.2,<1
numpy>=1.19.0
onnxruntime~=1.12.1
pytorch-lightning~=1.7.0
torch~=1.11.0
```

It is tested on Ununtu 18.04.

Thank you everyone, who advised and commented!
It is a strange the author's deep ignorance....

**Comment by msftcangoblowm** [comment]
@methosmythos

> Hi,
> 
> it seems the issue about espeak-phonemizer is already raised here: [rhasspy/espeak-phonemizer#1](https://github.com/rhasspy/espeak-phonemizer/issues/1) As far as I can tell it has to do with the python version. In my case it seems to work with python 3.8.10, but not with 3.8.12 and higher.

The issue is **NOT** affected by python version. One commenter mentioned python version. That commenter was wrong, but the thought persists. Like a mind-virus

espeak-phonemizer uses ctypes to make a call to the C library, libspeak-ng. The call needs to set the return type, .restype

Submitted a pull request which fixes this issue [#3](https://github.com/rhasspy/espeak-phonemizer/pull/3)

.... waiting for the author to review the pull request. It's one function!

**Comment by msftcangoblowm** [comment]
@alisherski 

> > It seems the "Segmentation fault (core dumped)" error is avoided, if we install older version of espeak_phonemizer.
> > `pip install espeak_phonemizer==1.1.0`
> 

> Thanks! I encountered the same error and can confirm that downgrading prevents segmentation fault. Launched preprocessing step, waiting for its finish.

Downgrading is a cope or at best a temporary solution.

**Just fix the call** (to libespeak-ng C library). Specify the return type....done!

python coders shouldn't be intimidated by calls to C libraries using ctypes python module

I'm sorry Piper users had to track down this issue to discover it originates in the upstream. And the upstream knew about this issue [#1](https://github.com/rhasspy/espeak-phonemizer/issues/1) and [how to fix it](https://github.com/rhasspy/espeak-phonemizer/issues/1#issuecomment-1510755337)

And there is a (pull request/) fix --> [#3](https://github.com/rhasspy/espeak-phonemizer/pull/3)

And all the author needs to do is review the pull request

**Comment by msftcangoblowm** [comment]
> Now I can say that the problem was fixed, if `src/python/requirements.txt` is edited such way:
> 
> ```
> cython>=0.29.0,<1
> espeak-phonemizer~=1.1.0
> librosa>=0.9.2,<1
> numpy>=1.19.0
> onnxruntime~=1.12.1
> pytorch-lightning~=1.7.0
> torch~=1.11.0
> ```
> 
> It is tested on Ununtu 18.04.
> 
> Thank you everyone, who advised and commented! It is a strange the author's deep ignorance....

tl;dr
python packaging poor. Make entrypoints; no bash scripts

espeak-phonemizer **AND** piper needs a rewrite; requires python packaging know-how

Both espeak-phonemizer **AND** piper, currently, are written as a cli utility, rather than as a library that also has a cli entrypoint. 

pyproject.toml now recommends **NOT** including bash scripts within python packages! So what is with all these bash scripts?! Remove the bash scripts and minimize the cli entrypoint code and what is left?

- espeak-phonemizer
One module ... one!

- piper --> run python
One module ... one!

- phonemizer2ids
One module ... one!

- piper --> trainer
Lots of modules

The entrypoint code (__main__.py) should do **only** arg parsing, but then should make calls to the module (__init__.py). The problem with code within __main__.py is **it's untestable**. Putting that code within a class and moving it to (__init__.py) would make it testable.

- Make entrypoints, not bash scripts!
- Replace setup.py with setup.cfg and pyproject.toml
- Uses package pytest, but tests are written as unittest. Write them as pytest
- Make Makefile great again!
- What is *typing.TYPE_CHECKING*
- What is *os.fspath* rather than str(some path)
- What is type[PurePath] rather than Path
- What is issubclass(type(some path), PurePath)
